src/perform_step/adams_bashforth_moulton_perform_step.jl

```diff
@@ -105,19 +119,21 @@ end
 
 @muladd function perform_step!(integrator,cache::ABM32Cache,repeat_step=false)
   @unpack t,dt,uprev,u,f,p = integrator
-  @unpack tmp,fsalfirst,k2,k3,ralk2,u2,k = cache
+  @unpack tmp,fsalfirst,k2,k3,ralk2,k = cache
   k1 = integrator.fsalfirst
-  cnt = length(integrator.sol.t)
+  cnt = integrator.iter
   if cnt == 1
-    @. ttmp = t + (2/3)*dt
-    f(ralk2, uprev + (2/3)*dt*k1, p, ttmp)
-    @. u = uprev + (dt/4)*(k1 + 3*ralk2)
+    ttmp = t + (2/3)*dt
+    @. tmp = uprev + (2/3)*dt*k1
+    f(ralk2, tmp, p, ttmp)
+    @. u = uprev + (dt/4)*(k1 + 3*ralk2)       #Ralston Method
+    cache.k2 = copy(k1)
   else
-    perform_step!(integrator,integrator.cache.onestep_cache)
-    u2 = integrator.sol(t-dt)
-    k2 = integrator.fsallast
-    f(k3, u2, p, t-dt)
-    @. u = uprev + (dt/12)*(5*k2 + 8*k1 - k3)
+    perform_step!(integrator, AB3Cache(u,uprev,fsalfirst,k2,k3,ralk2,k,tmp))
+    k = integrator.fsallast
+    @. u = uprev + (dt/12)*(5*k + 8*k1 - k2)
+    cache.k3 = copy(k2)
```

Comment:
For these you don't need to copy. Instead, a 2-way pointer rotation is fine. `cache.k2,cache.k3 = k3,k2`. I'm not sure you can do that all in one line but you get the point.  Then you need to move `k` into `k2` via `.=`.

Reply:
@ChrisRackauckas, By doing `cache.k2,cache.k3 = k1,k2` `chache.k2` refer to `k1` and  if `k1` changes then `cache.k2` will also change, here value should not change.

Reply:
I didn't suggest `cache.k2,cache.k3 = k1,k2` for that reason. Note that I mentioned that you can swap the last two pointers, but always need to `.=` for the `k` (`k1`) part.

Reply:
I got it.